src/include/daos/mem.h

```diff
@@ -632,4 +632,132 @@ umem_tx_add_callback(struct umem_instance *umm, struct umem_tx_stage_data *txd,
 	D_ASSERT(umm->umm_ops->mo_tx_add_callback != NULL);
 	return umm->umm_ops->mo_tx_add_callback(umm, txd, stage, cb, data);
 }
+
+/*********************************************************************************/
+
+typedef uint64_t	daos_addr_t;
+
+/** Describing a storage region for I/O */
+struct umem_store_region {
+	/** start offset of the region */
+	daos_addr_t	sr_addr;
+	/** size of the region */
+	daos_size_t	sr_size;
+};
+
+/** I/O descripter, it can include arbitrary number of storage regions */
+struct umem_store_iod {
+	/* number of regions */
+	int				 io_nr;
+	/** embedded one for single region case */
+	struct umem_store_region	 io_region;
+	struct umem_store_region	*io_regions;
+};
+
+/* Type of memory actions */
+enum {
+	UMEM_ACT_NOOP			= 0,
+	/** copy appended payload to specified storage address */
+	UMEM_ACT_COPY,
+	/** copy payload addressed by @ptr to specified storage address */
+	UMEM_ACT_COPY_PTR,
+	/** assign 8/16/32/64 bits integer to specified storage address */
+	UMEM_ACT_ASSIGN,
+	/** move specified bytes from source address to destination address */
+	UMEM_ACT_MOVE,
+	/** memset a region with specified value */
+	UMEM_ACT_SET,
+	/** set the specified bit in bitmap */
+	UMEM_ACT_SET_BITS,
+	/** unset the specified bit in bitmap */
+	UMEM_ACT_CLR_BITS,
+	/** it's checksum of the specified address */
+	UMEM_ACT_CSUM,
+};
+
+#define UMEM_ACT_PAYLOAD		24
+/**
+ * Memory operations for redo/undo.
+ * 16 bytes for bit operation (set/clr) and integer assignment, 32+ bytes for other operations.
+ */
+
+struct umem_action {
+	uint16_t			ac_opc;
+	union {
+		struct {
+			uint64_t		addr;
+			uint64_t		size;
+			uint8_t			payload[0];
+		} ac_copy;	/**< copy payload from @payload to @addr */
+		struct {
+			uint64_t		addr;
+			uint64_t		size;
+			uint64_t		ptr;
+		} ac_copy_ptr;	/**< copy payload from @ptr to @addr */
+		struct {
+			uint16_t		size;
+			uint32_t		val;
+			uint64_t		addr;
+		} ac_assign;	/**< assign integer to @addr, int64 should use ac_copy */
```

Comment:
the comment here and "uint32_t		val;" looks mismatch with above L664 "assign 8/16/32/64 bits integer to specified storage address"

Reply:
right...

Reply:
I guess it doesn't matter because it's a union and not in durable format, but I generally prefer fields to be largest to smallest so any padding is at the end.

Reply:
in WAL we will have many actions stored in a.4K page, each action may have different size, we cannot determine the size of action by the union size, instead we should compute the size by its type and the corresponding union member. 
I believe in each WAL transaction, we will have a few "assign" actions and each one can save 16-byte.